library/core/src/ptr/non_null.rs

```diff
@@ -1638,7 +1643,8 @@ impl<T> NonNull<[T]> {
     #[unstable(feature = "ptr_as_uninit", issue = "75402")]
     #[requires(self.as_ptr().cast::<T>().align_offset(core::mem::align_of::<T>()) == 0)] // Ensure the pointer is properly aligned
     #[requires(self.len().checked_mul(core::mem::size_of::<T>()).is_some() && self.len() * core::mem::size_of::<T>() <= isize::MAX as usize)] // Ensure the slice size does not exceed isize::MAX
-    #[requires(kani::mem::same_allocation(self.as_ptr() as *const(), self.as_ptr().byte_add(self.len() * core::mem::size_of::<T>()) as *const()))] // Ensure the slice is contained within a single allocation
+    #[requires(self.as_ptr().addr().checked_add(self.len() * core::mem::size_of::<T>()).is_some())]
+    #[requires(ub_checks::same_allocation(self.as_ptr() as *const(), self.as_ptr().wrapping_byte_add(self.len() * core::mem::size_of::<T>()) as *const()))] // Ensure the slice is contained within a single allocation
```

Comment:
Nit pick: this line is getting very long (and rustfmt doesn't yet know how to format contract lines), I'd suggest putting the comment above the requires clause so as to stick to the 100-chars line length limit.